File: src/subscriber.cpp
#include <dvl_a50_ros_driver/subscriber.hpp>

DVLA50Subscriber::DVLA50Subscriber() : Node("dvl_a50_subscriber") {
    raw_sub_ = this->create_subscription<std_msgs::msg::String>(
        "dvl/json_data", 10,
        std::bind(&DVLA50Subscriber::callbackRAW, this, std::placeholders::_1));
    dvl_sub_ = this->create_subscription<dvl_a50_ros_driver::msg::DVL>(
        "dvl/data", 10,
        std::bind(&DVLA50Subscriber::callback, this, std::placeholders::_1));
}

void DVLA50Subscriber::callbackRAW(const std_msgs::msg::String::SharedPtr msg) {
    RCLCPP_INFO(this->get_logger(), "Data received: %s", msg->data.c_str());
}

void DVLA50Subscriber::callback(const dvl_a50_ros_driver::msg::DVL::SharedPtr msg) {
    RCLCPP_INFO(this->get_logger(), "Time received: %f", msg->time);
}

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto subscriber = std::make_shared<DVLA50Subscriber>();
    rclcpp::spin(subscriber);
    rclcpp::shutdown();
    return 0;
}

File: scripts/subscriber.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from dvl_a50_ros_driver.msg import DVL
from dvl_a50_ros_driver.msg import DVLBeam


class DVLA50Subscriber(Node):
    def __init__(self):
        super().__init__('dvl_a50_subscriber')
        
        # Create subscriptions
        self.raw_subscription = self.create_subscription(
            String,
            'dvl/json_data',
            self.callbackRAW,
            10)
        self.dvl_subscription = self.create_subscription(
            DVL,
            'dvl/data',
            self.callback,
            10)
    
    def callbackRAW(self, data):
        self.get_logger().info(f'Data received: {data.data}')
    
    def callback(self, data):
        self.get_logger().info(f'Time received: {data.time}')


def main(args=None):
    rclpy.init(args=args)
    subscriber = DVLA50Subscriber()
    
    try:
        rclpy.spin(subscriber)
    except KeyboardInterrupt:
        pass
    finally:
        subscriber.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
